pkg/apis/eventing/struct_validation.go

```diff
@@ -0,0 +1,55 @@
+/*
+Copyright 2020 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package eventing
+
+import (
+	"reflect"
+)
+
+// HasMultipleSetFields checks if the struct has more than one field with a non-zero value
+func HasMultipleSetFields(o interface{}) bool {
+	val := reflect.ValueOf(o)
+	fieldSet := false
+	for i := 0; i < val.NumField(); i++ {
+		f := val.Field(i)
+		if !IsEmptyField(f) {
+			if !fieldSet {
+				fieldSet = true
+			} else {
+				return true
+			}
+		}
+	}
+	return false
+}
```

Comment:
Do we need this piece of generic and untyped logic even though we only use this function in one place and for only one type now?

I'd prefer generalizing it later if needed.

Reply:
I agree it's advised to generalize when needed, but do you have another approach to validate that a single struct field is set other than a bunch of "if filter.Exact.... if filter.Prefex...etc" where you iterate on each single field?

Reply:
I don't feel this is more readable than the proposed solution IMHO. The solution does exactly what is needed. We need to check at run time how many fields of a struct is set to none-zero, what is that if not straight forward reflection? It's typed but it lacks a safety check though (I'll add a check that it's a struct and return an error otherwise) , reusable and generic. 

It's also not in the hot path of any data-plane to worry about the reflection performance overhead, this is only control-plane logic.

I'm honestly trying to understand the actual concerns, is it just a preference?

Reply:
It's not readable and a swith like this:
```
	switch v.Kind() {
	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
		return v.Int() == 0
	case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64:
		return v.Uint() == 0
	case reflect.String:
		return v.String() == ""
	case reflect.Ptr, reflect.Slice, reflect.Map, reflect.Interface, reflect.Chan:
		return v.IsNil()
	case reflect.Bool:
		return !v.Bool()
	}
	return false
```
isn't any different than a bunch of "if filter.Exact.... if filter.Prefex...etc"

Reply:
No performance concern here.

Reply:
> It's not readable and a switch like this:
> 

I'd honestly push hard against this argument, but again maybe it's MY preference and I'm having a hard time seeing it.

I can also see the disadvantage of losing compile time validation in general, but that's when "overusing" reflection and this case is IMHO a very narrowed down use case of reflection that IMHO is considered the safest use of reflection, "read-only inspection of a datatype" . 

Anyway, lemme know how you'd like to proceed I'm happy to change it and move forward with the PR

> ```go
> 	switch v.Kind() {
> 	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
> 		return v.Int() == 0
> 	case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64:
> 		return v.Uint() == 0
> 	case reflect.String:
> 		return v.String() == ""
> 	case reflect.Ptr, reflect.Slice, reflect.Map, reflect.Interface, reflect.Chan:
> 		return v.IsNil()
> 	case reflect.Bool:
> 		return !v.Bool()
> 	}
> 	return false
> ```
> 
> isn't any different than a bunch of "if filter.Exact.... if filter.Prefex...etc"

Exactly same opinion as the above. If the language had a `reflect.NonZeroFields(*interface{})` I'd have definitely used it and it'd have probably looked something like that switch inside. However, I can tell you're strongly against it and I'd love to move forward.

Reply:
One of the 2 alternative approaches looks simpler to me.

Reply:
I have the same concern regarding the `interface{}` parameter and the overly generic code.

This function don't accept _any struct_, it accepts `SubscriptionsAPIFilter` structs and _only_ that type.

By tailoring your code for that case, you make explicit what you expect as input and what you are going to check. As a result, the intent is clear and your code is easier to read and maintain. I don't see this as a personal preference, I see this as a generally well accepted practice.

Reply:
> This function don't accept any struct, it accepts SubscriptionsAPIFilter structs and only that type.

I'm not sure I got you correctly, but this function **does** accept and works for any struct, hence the exported, generic name.

I'll refactor it though, as I said I wanna move this PR forward.

Reply:
> this function does accept and works with any struct

The function signature yes.
In the usage, no.